bugfix: after using curl_multi_fdset(), selcet() timeout minimum is recommended to be set 100ms, rather than NULL, which may cause current thread always waiting if 'maxfd' equal -1 on some platform.

yanxiuzi opened the pull request:
see https://github.com/curl/curl/blob/5ccddf64398c1186deb5769dac086d738e150e09/docs/examples/multi-legacy.c

/*On success the value of maxfd is guaranteed to be >= -1. We call
       select(maxfd + 1, ...); specially in case of (maxfd == -1) there are
       no fds ready yet so we call select(0, ...) --or Sleep() on Windows--
       to sleep 100ms, which is the minimum suggested value in the
       curl_multi_fdset() doc.*/

balamurugana reviewed (changes requested):
What exactly the issue instead of pointing to curl example?

yanxiuzi commented:
> What exactly is the issue instead of pointing to curl example?

When I deployed the project to l4t, the initialization program blocked at `s3client_->BucketExists(args)`.
After debuging,  it is determined to be that blocked at `selcet()` while `maxfd==-1` either `timeout` is NULL, for there are no curl fd ready yet.

yanxiuzi commented:
> > What exactly is the issue instead of pointing to curl example?
> 
> When I deployed my project to l4t, the program blocked at `s3client_->BucketExists(args)` when initializing. After debuging, it is determined to be that blocked at `selcet()` while `maxfd==-1` either `timeout` is NULL, for there are no curl fd ready yet.

After applying this patch, everything is now working fine.

balamurugana commented:
> > > What exactly is the issue instead of pointing to curl example?
> > 
> > 
> > When I deployed my project to l4t, the program blocked at `s3client_->BucketExists(args)` when initializing. After debuging, it is determined to be that blocked at `selcet()` while `maxfd==-1` either `timeout` is NULL, for there are no curl fd ready yet.
> 
> After applying this patch, everything is now working fine.

It seems a curl bug. You suppose to open an issue there. There is no point in adding workaround here.

balamurugana commented:
CI runs functional tests which succeeds. If there is a problem in curlpp usage, it should fail in CI. We need a concrete proof of failure.

yanxiuzi commented:
> CI runs functional tests which succeeds. If there is a problem in curlpp usage, it should fail in CI. We need a concrete proof of failure.

My programe using this library worked fine on multiple intel x86 linux servers, before deploying to armv8 l4t today. However, the problem is almost certain to occur during several debugging sessions today. I have fixed the problem through the fork repository and packaged the new binary so file, so the time it takes for me to reproduce the previous situation is similar to that of locating this problem, which is more troublesome. For this patch, I suggest dev team to make a test,  although it works well for me now.   (armv8,  ubuntu 18.04  l4t, g++-8, dynamic link ssl, curl and curlpp)

yanxiuzi commented:
> > CI runs functional tests which succeeds. If there is a problem in curlpp usage, it should fail in CI. We need a concrete proof of failure.
> 
> My programe using this library worked fine on multiple intel x86 linux servers, before deploying to armv8 l4t today. However, the problem is almost certain to occur during several debugging sessions today. I have fixed the problem through the fork repository and packaged the new binary so file, so the time it takes for me to reproduce the previous situation is similar to that of locating this problem, which is more troublesome. For this patch, I suggest dev team to make a test, although it works well for me now. (armv8, ubuntu 18.04 l4t, g++-8, dynamic link ssl, curl and curlpp)

> > CI runs functional tests which succeeds. If there is a problem in curlpp usage, it should fail in CI. We need a concrete proof of failure.
> 
> My programe using this library worked fine on multiple intel x86 linux servers, before deploying to armv8 l4t today. However, the problem is almost certain to occur during several debugging sessions today. I have fixed the problem through the fork repository and packaged the new binary so file, so the time it takes for me to reproduce the previous situation is similar to that of locating this problem, which is more troublesome. For this patch, I suggest dev team to make a test, although it works well for me now. (armv8, ubuntu 18.04 l4t, g++-8, dynamic link ssl, curl and curlpp)


It is also possible that the libcurl-dev and libcurlpp-dev of the system apt repository version too low, I dont remember their version at this time. Any way, this patch make library more adaptable.

balamurugana commented:
As I told earlier we cannot workaround. My recommendation is to fix it locally.
